subsys/fs/littlefs_fs.c

```diff
@@ -737,90 +760,109 @@ static int littlefs_mount(struct fs_mount_t *mountp)
+		block_size = get_block_size_blk((char *)fs->backend);
 	}
 
-	lfs_size_t lookahead_size = lcp->lookahead_size;
+	if (IS_ENABLED(CONFIG_FS_LITTLEFS_FMP_DEV) && fmp_dev) {
+		block_size = get_block_size_fmp((struct flash_area *)fs->backend);
+	}
 
-	if (lookahead_size == 0) {
-		lookahead_size = CONFIG_FS_LITTLEFS_LOOKAHEAD_SIZE;
+	if (block_size == 0U) {
+		LOG_ERR("Failed to retrieve block size");
+		ret = -EINVAL;
+		goto out;
 	}
 
-	/* No, you don't get to override this. */
-	lfs_size_t block_count;
+	lfs_size_t block_count = 0U;
 
 	if (IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && block_dev) {
-		ret = disk_access_ioctl((char *) fs->backend,
-					DISK_IOCTL_GET_SECTOR_COUNT,
-					&block_count);
-		if (ret < 0) {
-			LOG_ERR("Unable to get sector count!");
+		block_count = get_block_cnt_blk((char *)fs->backend);
+	}
+
+	if (IS_ENABLED(CONFIG_FS_LITTLEFS_FMP_DEV) && fmp_dev) {
+		size_t size = ((struct flash_area *)fs->backend)->fa_size;
+
+		if ((size % block_size) != 0U) {
+			LOG_ERR("Flash area size not a multiple of blocksize");
 			ret = -EINVAL;
 			goto out;
 		}
+		block_count = size / block_size;
+	}
+
+	if (block_count == 0U) {
+		LOG_ERR("Failed to retrieve block count");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* recalculate block_size and block_count if override block_size was
+	 * provided
+	 */
+	uint32_t scale = lcp->block_size / block_size;
+
+	if ((scale != 0U) && (scale * block_size == lcp->block_size) &&
+	    (block_count % scale == 0U)) {
+		block_count /= scale;
+		block_size = lcp->block_size;
+	}
+
+	if (IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && block_dev) {
 		LOG_INF("FS at %s: is %u 0x%x-byte blocks with %u cycle",
-			(char *) fs->backend, block_count, block_size,
+			(char *)fs->backend, block_count, block_size,
 			block_cycles);
-	} else {
-		block_count = ((struct flash_area *)fs->backend)->fa_size
-			/ block_size;
+	}
+
+	if (IS_ENABLED(CONFIG_FS_LITTLEFS_FMP_DEV) && fmp_dev) {
 		const struct device *dev =
 			flash_area_get_device((struct flash_area *)fs->backend);
 		LOG_INF("FS at %s:0x%x is %u 0x%x-byte blocks with %u cycle",
 			log_strdup(dev->name),
 			(uint32_t)((struct flash_area *)fs->backend)->fa_off,
 			block_count, block_size, block_cycles);
-		LOG_INF("sizes: rd %u ; pr %u ; ca %u ; la %u",
-			read_size, prog_size, cache_size, lookahead_size);
 	}
 
-	__ASSERT_NO_MSG(prog_size != 0);
-	__ASSERT_NO_MSG(read_size != 0);
-	__ASSERT_NO_MSG(cache_size != 0);
-	__ASSERT_NO_MSG(block_size != 0);
+	LOG_INF("sizes: rd %u ; pr %u ; ca %u ; la %u",	read_size, prog_size,
+		cache_size, lookahead_size);
 
-	__ASSERT((block_size % prog_size) == 0,
-		 "erase size must be multiple of write size");
-	__ASSERT((block_size % cache_size) == 0,
-		 "cache size incompatible with block size");
+	if ((block_size % prog_size) != 0U) {
+		LOG_ERR("Erase size must be multiple of write size");
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if ((block_size % cache_size) != 0U) {
```

Comment:
In my case (for block device test setup) it is (512B % 2048) which is 512 and the code exits early with error.

Please adjust rest of those checks accordingly (as I also see some errors for Flash memory)

Reply:
Right, the cache size needs to be a factor of the block size, I don't know what the optimal value would be for sdcard (and the combination of sdcard and flash). I will set it as default to 512 for sdcard support.